src/Logger-PREVAC_TM1x/main.py

```diff
@@ -0,0 +1,388 @@
+            self.initial_frequency,
+            self.material_density,
+            self.impedance_ratio,
+        )
+        thickness *= self.tooling_factor
+
+        rate = self.calculate_rate(thickness)
+        return [frequency, thickness, rate]
+
+    """ Device Functions """
+
+    def get_frequency(self) -> float:
+        """Request current frequency in Hz.
+
+        response[:4]    frequency in Hz with 0.01 Hz resolution
+        response[4]     1 microbalance correction connected, 2 not
+        response[5]     measurements are numbered from 0 to 255
+        response[6:7]   duration of the measurement in ms
+        response[7]     13 for TM13, 14 for TM14
+        """
+        command = 0x53
+        data = [self.sample_rate, 0, 0, 0]
+
+        self.send_dataframe(command, data)
+        response = self.get_dataframe()
+
+        frequency_bytes = response[:4].encode("latin-1")
+        frequency = int.from_bytes(frequency_bytes, "big")
+
+        return frequency / 100
+
+    @staticmethod
+    def calculate_thickness(
+        frequency: float,
+        initial_frequency: float,
+        density_material: float,
+        impedance_ratio: float,
+    ) -> float:
+        """Calculate the thickness from the frequency using Sauerbreys equation."""
+        freq_constant = 1.668e5  # Hz * cm
+        density_quartz = 2.648  # g/cm^3
+
+        normalization = (
+            density_quartz * freq_constant / (np.pi * density_material * impedance_ratio * initial_frequency)
+        )
+        ratio = np.arctan(impedance_ratio * np.tan(np.pi * (frequency - initial_frequency) / frequency))
+
+        # TODO: Check unit
+        # Convert from cm to nm
+        return normalization * ratio * 1e7 * -1
+
+    def reset_thickness(self) -> None:
+        """Reset thickness by updating initial frequency."""
+        self.initial_frequency = self.get_frequency()
+
+    def calculate_rate(self, thickness: float) -> float:
+        """Calculate the rate of the thickness change in A/s."""
+        # TODO: Check if the device measures without given command at every sample rate interval.
+        # If yes, the rate should be calculated using the sample rate instead of the time stamp if time stamp is smaller
+        time_stamp = time.time()
+
+        # First measurement or after reset_thickness
+        if self.previous_thickness == 0 or self.previous_time == 0:
+            rate = 0
+            self.previous_time = time_stamp
+
+        # Time stamp is the same as the previous time stamp, would give division by 0 error.
+        elif time_stamp == self.previous_time:  # noqa: SIM114
+            rate = self.previous_rate
+
+        # If calculate_rate is called faster than the sampling rate of the device, the thickness does not change.
+        # This would lead to a rate of 0. To avoid this, the rate is set to the previous rate.
+        elif thickness == self.previous_thickness and time_stamp - self.previous_time < 1 / self.sample_rate:
```

Comment:
> `1 / self.sample_rate`
- From looking at `self.sample_rate_dict`, the variable `self.sample_rate` does not contain a frequency in Hz, but rather some index, or key? In that case, `1 / self.sample_rate` will produce unexpected results
- Not strictly an issue in Python 3, but in other languages or even Python 2, a division of two integers will always result in an integer, i.e. 1/4 = 0. To make the division yield a float, one of the arguments needs to be a float, e.g. by writing 1.0 / variable. As I said, it's not a problem in Python 3, as it will always give a float (e.g. 1/1 = 1.0), but I still prefer to make it explicit that a division of floats is intended.